File: C++/LabWork#3/TrapezoidOperations.cpp
#include "TrapezoidOperations.h"

TrapezoidOperations::TrapezoidOperations() {
  A = { 0, 0 };
  B = { 4, 0 };
  C = { 3, 2 };
  D = { 1, 2 };
}

TrapezoidOperations::TrapezoidOperations(Point a, Point b, Point c, Point d) : A(a), B(b), C(c), D(d) {}

TrapezoidOperations::TrapezoidOperations(const TrapezoidOperations& other) : A(other.A), B(other.B), C(other.C), D(other.D) {}

double TrapezoidOperations::getArea() const {
  double h = std::abs(C.Y - A.Y);
  double ab = std::abs(B.X - A.X);
  double cd = std::abs(C.X - D.X);
  return 0.5 * (ab + cd) * h;
}

double TrapezoidOperations::getPerimeter() const {
  return distanceBetweenPoints(A, B) + distanceBetweenPoints(B, C) + distanceBetweenPoints(C, D) + distanceBetweenPoints(D, A);
}

double TrapezoidOperations::distanceBetweenPoints(const Point& p1, const Point& p2) const {
  return std::sqrt(std::pow(p2.X - p1.X, 2) + std::pow(p2.Y - p1.Y, 2));
}

TrapezoidOperations TrapezoidOperations::operator-() const {
  return TrapezoidOperations(
    { A.X - 3, A.Y - 3 },
    { B.X - 3, B.Y - 3 },
    { C.X - 3, C.Y - 3 },
    { D.X - 3, D.Y - 3 }
  );
}

TrapezoidOperations TrapezoidOperations::operator*(const TrapezoidOperations& other) const {
  return TrapezoidOperations(
    { A.X * other.A.X, A.Y * other.A.Y },
    { B.X * other.B.X, B.Y * other.B.Y },
    { C.X * other.C.X, C.Y * other.C.Y },
    { D.X * other.D.X, D.Y * other.D.Y }
  );
}

void TrapezoidOperations::printCoordinates() const {
  std::cout << "A: (" << A.X << ", " << A.Y << ")" << std::endl;
  std::cout << "B: (" << B.X << ", " << B.Y << ")" << std::endl;
  std::cout << "C: (" << C.X << ", " << C.Y << ")" << std::endl;
  std::cout << "D: (" << D.X << ", " << D.Y << ")" << std::endl;
}

File: C#/LabWork#3/TrapezoidOperations.cs
using System;

public class TrapezoidOperations
{
  private (double X, double Y) _a;
  private (double X, double Y) _b;
  private (double X, double Y) _c;
  private (double X, double Y) _d;

  public TrapezoidOperations()
  {
    _a = (0, 0);
    _b = (4, 0);
    _c = (3, 2);
    _d = (1, 2);
  }

  public TrapezoidOperations((double, double) a, (double, double) b, (double, double) c, (double, double) d)
  {
    _a = a;
    _b = b;
    _c = c;
    _d = d;
  }

  public TrapezoidOperations(TrapezoidOperations otherTr)
  {
    _a = otherTr._a;
    _b = otherTr._b;
    _c = otherTr._c;
    _d = otherTr._d;
  }

  public double GetArea()
  {
    double h = Math.Abs(_c.Y - _a.Y);
    double ab = Math.Abs(_b.X - _a.X);
    double cd = Math.Abs(_c.X - _d.X);
    return 0.5 * (ab + cd) * h;
  }

  public double GetPerimeter()
  {
    return DistanceBetweenPoints(_a, _b) + DistanceBetweenPoints(_b, _c) + DistanceBetweenPoints(_c, _d) + DistanceBetweenPoints(_d, _a);
  }

  private double DistanceBetweenPoints((double X, double Y) p1, (double X, double Y) p2)
  {
    return Math.Sqrt(Math.Pow(p2.X - p1.X, 2) + Math.Pow(p2.Y - p1.Y, 2));
  }

  public TrapezoidOperations operator -(TrapezoidOperations t)
  {
    return new TrapezoidOperations(
        (t._a.X - 3, t._a.Y - 3),
        (t._b.X - 3, t._b.Y - 3),
        (t._c.X - 3, t._c.Y - 3),
        (t._d.X - 3, t._d.Y - 3)
    );
  }

  public TrapezoidOperations operator *(TrapezoidOperations t1, TrapezoidOperations t2)
  {
    return new TrapezoidOperations(
        (t1._a.X * t2._a.X, t1._a.Y * t2._a.Y),
        (t1._b.X * t2._b.X, t1._b.Y * t2._b.Y),
        (t1._c.X * t2._c.X, t1._c.Y * t2._c.Y),
        (t1._d.X * t2._d.X, t1._d.Y * t2._d.Y)
    );
  }
  public void PrintCoordinates()
  {
    Console.WriteLine($"A: ({_a.X}, {_a.Y})");
    Console.WriteLine($"B: ({_b.X}, {_b.Y})");
    Console.WriteLine($"C: ({_c.X}, {_c.Y})");
    Console.WriteLine($"D: ({_d.X}, {_d.Y})");
  }
}
